deploy-agent/deployd/common/utils.py

```diff
@@ -101,36 +105,109 @@ def uptime():
 
-def run_prereqs(config):
-    # check if the puppet has finished or not
+def is_first_run(config):
+    env_status_file = config.get_env_status_fn()
+    return not os.path.exists(env_status_file)
+
+
+def check_prereqs(config):
+    """
+    Check prerequisites before deploy agent can run
+
+    :return: True all conditions meet else False
+    """
     if IS_PINTEREST:
         respect_puppet = config.respect_puppet()
-        puppet_file_path = config.get_puppet_file_path()
-        if respect_puppet and \
-           puppet_file_path is not None and \
-           not os.path.exists(puppet_file_path):
-            print("Waiting for first puppet run.")
-            sys.exit(0)
+        # check if the puppet has finished successfully or not
+        if respect_puppet:
+            puppet_state_file_path = config.get_puppet_state_file_path()
+            if puppet_state_file_path and (not os.path.exists(puppet_state_file_path)):
+                log.error("Waiting for first puppet run.")
+                return False
+            if not check_first_puppet_run_success(config):
+                log.error("First puppet run failed.")
+                return False
 
     ensure_dirs(config)
+    return True
+
+
+def get_puppet_exit_code(config):
+    """
+    Get puppet exit code from the corresponding file
+
+    :return: puppet exit code or 999 if file doesn't exist
+    """
+    puppet_exit_code_file = config.get_puppet_exit_code_file_path()
+    try:
+        with open(puppet_exit_code_file, "rt") as f:
+            exit_code = f.readline().strip()
+    except Exception as e:
+        log.warning(f"Could not read {puppet_exit_code_file} file: {e}")
+        exit_code = 999
+
+    return exit_code
+
+
+def load_puppet_summary(config):
+    """
+    Load last_run_summary yaml file, parse results
+
+    :return: returns a dict constructed from for the puppet summary file
+    """
+    summary_file = config.get_puppet_summary_file_path()
+    summary = {}
+    if not os.path.exists(summary_file):
+        log.warning(f"{summary_file} does not exist. This could be the first puppet run")
+        return summary
+
+    with open(summary_file) as f:
+        summary = yaml.safe_load(f)
+    return summary
+
+
+def check_first_puppet_run_success(config):
+    """
+    Check first puppet run success from exit code and last run summary
+
+    :return: returns True if success else False
+    """
+    if not is_first_run(config):
+        return True
+
+    puppet_exit_code = get_puppet_exit_code(config)
+    if puppet_exit_code in PUPPET_SUCCESS_EXIT_CODES:
+        return True
+
+    # If failed, double check with puppet last summary
+    puppet_summary = load_puppet_summary(config)
+    puppet_failures = puppet_summary.get('events', {}).get(
+        'failure', None) if puppet_summary else None
+    log.info(f"Puppet failures: {puppet_failures}")
+
+    if puppet_failures != 0:
+        send_statsboard_metric(name='deployd.first_puppet_failed', value=1,
```

Comment:
I think we can send both way not just failure, with a 'result' tag to differentiate.

Reply:
we don't want to send too many requests to statsboard using http calls, we also care for puppet failures only so sending both will not add much value and waste resource.